src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs

```diff
@@ -391,24 +360,101 @@ private void LowerBasicSwitch(DecisionTree.ByValue byValue)
                 }
 
                 var rewrittenSections = switchSections.ToImmutableAndFree();
+
                 MethodSymbol stringEquality = null;
-                if (byValue.Type.SpecialType == SpecialType.System_String)
+                switch (byValue.Type.SpecialType)
                 {
-                    LocalRewriter.EnsureStringHashFunction(rewrittenSections, _factory.Syntax);
-                    stringEquality = LocalRewriter.GetSpecialTypeMethod(_factory.Syntax, SpecialMember.System_String__op_Equality);
+                    case SpecialType.System_String:
+                        {
+                            LocalRewriter.EnsureStringHashFunction(rewrittenSections, _factory.Syntax);
+                            stringEquality = LocalRewriter.GetSpecialTypeMethod(_factory.Syntax, SpecialMember.System_String__op_Equality);
+                            goto case SpecialType.System_UInt32;
+                        }
+
+                    case SpecialType.System_Byte:
+                    case SpecialType.System_Char:
+                    case SpecialType.System_Int16:
+                    case SpecialType.System_Int32:
+                    case SpecialType.System_Int64:
+                    case SpecialType.System_SByte:
+                    case SpecialType.System_UInt16:
+                    case SpecialType.System_UInt32:
+                    case SpecialType.System_UInt64:
+                        {
+                            // Emit requires a constant target when there are no sections, so we accomodate that here.
+                            // CONSIDER: can we get better code generated by giving a constant target more often here,
+                            // e.g. when the switch expression is a constant?
+                            var constantTarget = rewrittenSections.IsEmpty ? noValueMatches : null;
+                            var switchStatement = new BoundSwitchStatement(
+                                _factory.Syntax, null, _factory.Convert(underlyingSwitchType, byValue.Expression),
+                                constantTarget,
+                                ImmutableArray<LocalSymbol>.Empty, ImmutableArray<LocalFunctionSymbol>.Empty,
+                                rewrittenSections, noValueMatches, stringEquality);
+                            _loweredDecisionTree.Add(switchStatement);
+                            // The bound switch statement implicitly defines the label noValueMatches at the end, so we do not add it explicitly.
+                            break;
+                        }
+
+                    default:
+                        {
+                            if (byValue.Type.TypeKind == TypeKind.Enum)
+                            {
+                                goto case SpecialType.System_Int32;
```

Comment:
I'll break out the float/double/decimal code handling. The primitive handling code is only three lines.

---
In reply to: [70701623](https://github.com/dotnet/roslyn/pull/12478#discussion_r70701623) [](ancestors = 70701623)